lib/psci/psci_private.h

```diff
@@ -27,18 +27,6 @@
 
 #define psci_dsbish()
 
-/*
- * On systems where participant CPUs are cache-coherent, we can use spinlocks
- * instead of bakery locks.
- */
-#define DEFINE_PSCI_LOCK(_name)		spinlock_t _name
-#define DECLARE_PSCI_LOCK(_name)	extern DEFINE_PSCI_LOCK(_name)
-
-#define psci_lock_get(non_cpu_pd_node)				\
-	spin_lock(&psci_locks[(non_cpu_pd_node)->lock_index])
-#define psci_lock_release(non_cpu_pd_node)			\
-	spin_unlock(&psci_locks[(non_cpu_pd_node)->lock_index])
-
```

Comment:
The HW_ASSISTED_COHERENCY flag is intended for use in Arm DynamIQ platforms (or similar), which are v8.2+. On these platforms, it is expected that the Compare and Swap (CAS) instruction is available and that this is used in the underlying spinlock implementation, by ensuring ARM_ARCH_MINOR is at least 1.

In any case, a great deal effort has gone into ensuring spinlocks will work correctly in the situations they are used in the codebase. It is highly undesirable to remove them in favour of much less efficient bakery locks, unless this is absolutely necessary. We need to understand the specific problem in your system (and try to fix that) rather than hitting performance in all systems.

Reply:
Then the flag is a misnomer, as we have HW assisted coherency but are not using v8.2+ cores. Perhaps we can break this into two flags, one that means what it says and deals with cache, and a new one that defines whether to use spinlocks vs bakery locks.

The comment above the code is also wrong, CPUs being cache-coherent does not guarantee we can safely use these spinlocks, it only removes one of many pitfalls described in the ARM ARM that will break them. So as above I think we need a new flag for using spinlocks, but for now this patch is correct in removing the lock type selection from under the scope of the unrelated flag HW_ASSISTED_COHERENCY.

Reply:
In our definition of HW_ASSISTED_COHERENCY, data caches can be enabled very early on boot (i.e. without requiring CCI or other platform programming) and remain enabled right up to final power off. This makes it possible for us to comply with the rules around safe usage of exclusive access instructions, which in turn means it's architecturally safe to use spinlocks. I don't know if your definition of HW asssisted coherency is the same. If you think we are not complying with the Arm ARM in some way, then please state the specific issue. 

The option of using CAS instead of exclusive acccess instructions is just an optimisation; it should work either way.